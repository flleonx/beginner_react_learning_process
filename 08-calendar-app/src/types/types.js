export const types = {
  uiOpenModal: "[ui] Open modal",
  uiCloseModal: "[ui] Close modal",

  calendarEventStartAddNew: "[calendarEvent] Start add new",
  calendarEventAddNew: "[calendarEvent] Add new",
  calendarEventSetActive: "[calendarEvent] Set active",
  calendarEventClearActive: "[calendarEvent] Clear active",
  calendarEventUpdate: "[calendarEvent] Update",
  calendarEventDelete: "[calendarEvent] Delete",
  calendarEventLoaded: "[calendarEvent] Loaded",
  calendarEventLogout: "[calendarEvent] Clean when logout",

  authCheckingFinish: "[auth] Checking login state finished",
  authStart: "[auth] Start login",
  authLogin: "[auth] Login",
  authStartRegister: "[auth] Start register",
  authStarRenew: "[auth] Start token renew",
  authLogout: "[auth] Logout",
};
