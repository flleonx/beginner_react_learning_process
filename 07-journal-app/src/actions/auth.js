import Swal from 'sweetalert2'
import {
  createUserWithEmailAndPassword,
  signInWithPopup,
  updateProfile,
  signInWithEmailAndPassword,
  signOut
} from 'firebase/auth';
import { auth, googleAuthProvider } from "../firebase/firebaseConfig";
import { types } from "../types/types";
import { finishLoading, startLoading } from './ui';
import { noteLogout } from './notes';

export const startLoginEmailPassword = ( email, password ) =>{
  return async (dispatch) => {
    try {
      dispatch( startLoading() );
      const { user } = await signInWithEmailAndPassword( auth, email, password );
      dispatch( login( user.uid, user.displayName ) );
      dispatch( finishLoading() );
    } catch (error) {
      dispatch( finishLoading() );
      Swal.fire("Error", error.message, "error");
      console.log(error);
    }
  };
};

export const startRegisterWithEmailPasswordName = (name, email, password) => {
  return async ( dispatch ) => {
    try {
      const { user } = await createUserWithEmailAndPassword( auth, email, password );
      await updateProfile( user, { displayName: name });
      dispatch( login(user.uid, user.displayName) );
    } catch (error) {
      Swal.fire("Error", error.message, "error");
      console.log(error);
    }
  };
};

export const startGoogleLogin = () => {
  return async ( dispatch ) => {
    const { user } = await signInWithPopup( auth, googleAuthProvider );
    dispatch( login( user.uid, user.displayName ));
  };
};

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName
  }
});

export const startLogout = () => {
  return async ( dispatch ) => {
    await signOut( auth );
    dispatch( logout() );
    dispatch( noteLogout() );
  };
};

export const logout = () => ({
  type: types.logout
});
