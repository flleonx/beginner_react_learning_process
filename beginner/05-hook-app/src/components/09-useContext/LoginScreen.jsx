import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const LoginScreen = () => {

  // 1. Get userContext's reference
  const { setUser } = useContext( UserContext );
  // 2. setUser
  const user = {
    id: 1234,
    name: 'flleonx'
  };

  return (
    <>
      <h1>LoginScreen</h1>
      <hr/>

      <button
        className="btn btn-primary"
        onClick={ () => setUser(user) }
      >
        Login
      </button>

    </>
  )
};
