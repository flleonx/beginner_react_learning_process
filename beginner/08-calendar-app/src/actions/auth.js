import Swal from "sweetalert2";

import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";
import { calendarEventLogout } from "./calendarEvent";

export const startLogin = (email, password) => {
  return async (dispatch) => {
    const resp = await fetchWithoutToken("auth", { email, password }, "POST");
    const body = await resp.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(login({
        uid: body.uid,
        name: body.name
      }));
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  };
};

const login = (user) => ({
  type: types.authLogin,
  payload: user
});

export const startRegister = (name, email, password) => {
  return async (dispatch) => {
    const resp = await fetchWithoutToken("auth/new", { name, email, password }, "POST");
    const body = await resp.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(login({
        uid: body.uid,
        name: body.name
      }));
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  };
};

export const startChecking = () => {
  return async (dispatch) => {
    const resp = await fetchWithToken("auth/renew");
    const body = await resp.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(login({
        uid: body.uid,
        name: body.name
      }));
    } else {
      dispatch(checkingFinish());
    }
  };
};

const checkingFinish = () => ({ type: types.authCheckingFinish });

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(calendarEventLogout());
    dispatch(logout());
  };
};

const logout = () => ({ type: types.authLogout })
