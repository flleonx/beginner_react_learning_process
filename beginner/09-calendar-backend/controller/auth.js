const { response } = require("express");
const bcrypt = require("bcryptjs");
const { generateJWT } = require("../helpers/jwt");

const User = require("../models/User");

const createUser = async (req, res = response) => {

  const { email, password } = req.body;

  try {

    let user = await User.findOne({ email });

    if (user) {
      return res.status(400).json({
        ok: false,
        msg: "There is already a user with that email"
      });
    }

    user = new User(req.body);

    // Password encryption
    const salt = bcrypt.genSaltSync();
    user.password = bcrypt.hashSync(password, salt);

    await user.save();

    // Generate JWT
    const token = await generateJWT(user.id, user.name);

    return res.status(201).json({
      ok: true,
      uid: user.id,
      name: user.name,
      token
    });

  } catch (error) {
    console.log(error);
    return res.status(500).json({
      ok: false,
      msg: "Please talk with the admin",
    });
  }
};

const loginUser = async (req, res = response) => {

  try {

    const { email, password } = req.body;

    const user = await User.findOne({ email });

    // Use a more generic msg
    if (!user) {
      return res.status(400).json({
        ok: false,
        msg: "There is no user with that email"
      });
    };

    // Validate password
    const validPassword = bcrypt.compareSync(password, user.password);

    if (!validPassword) {
      return res.status(400).json({
        ok: false,
        msg: "Wrong password"
      });
    };

    // Generate JWT
    const token = await generateJWT(user.id, user.name);

    return res.json({
      ok: true,
      uid: user.id,
      name: user.name,
      token
    });

  } catch (error) {
    console.log(error);
    return res.status(500).json({
      ok: false,
      msg: "Please talk with the admin",
    });
  }

};

const renewToken = async (req, res = response) => {

  const { uid, name } = req;

  // Generate JWT
  const token = await generateJWT(uid, name);

  return res.status(200).json({
    ok: true,
    uid,
    name,
    token
  });

};

module.exports = {
  createUser,
  loginUser,
  renewToken,
};
