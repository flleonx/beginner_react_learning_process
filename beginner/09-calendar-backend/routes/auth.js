/*
  User routes / Auth
  host + /api/auth
*/
const { Router } = require("express");

const { createUser, loginUser, renewToken } = require("../controller/auth");
const { createUserSchema, loginUserSchema } = require("../models/joi-schemas");
const { loadSchema, fieldsValidator } = require("../middlewares/fields-validator");
const { jwtValidator } = require("../middlewares/jwt-validator");

const auth = Router();

auth.post("/", [
  loadSchema(loginUserSchema),
  fieldsValidator
], loginUser);

auth.post("/new", [
  loadSchema(createUserSchema),
  fieldsValidator
],createUser);

auth.get("/renew", [
  jwtValidator
], renewToken);

module.exports = auth;
