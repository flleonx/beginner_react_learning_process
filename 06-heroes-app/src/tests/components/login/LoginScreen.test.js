import { mount } from "enzyme";
import { MemoryRouter, Routes, Route } from "react-router-dom";

import { AuthContext } from "../../../auth/authContext";
import { LoginScreen } from "../../../components/login/LoginScreen";
import { types } from "../../../types/types";

const mockDispatch = jest.fn();
const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

describe('Tests on <LoginComponent />', () => {

  const action = {
    type: types.login,
    payload: {
      name: "flleonx",
    }
  };

  const contextValue = {
    user: {
      logged: false
    },
    dispatch: mockDispatch
  };

  const wrapper = mount(
    <AuthContext.Provider value={ contextValue }>
      <MemoryRouter initialEntries={["/login"]}>
        <Routes>
          <Route path="/login" element={<LoginScreen />} />
        </Routes>
      </MemoryRouter>
    </AuthContext.Provider>
  );

  test('It should display <LoginComponent /> properly', () => {
    expect( wrapper ).toMatchSnapshot();
  });

  test('It should call dispatch and navigate', () => {
    const button = wrapper.find("button");
    button.simulate("click");
    expect( mockDispatch ).toHaveBeenCalledWith( action );
    expect( mockNavigate ).toHaveBeenCalledWith("/marvel", { replace: true });

    localStorage.setItem("lastPath", "/dc");
    button.simulate("click");
    expect( mockNavigate ).toHaveBeenCalledWith("/dc", { replace: true });
  });

});
