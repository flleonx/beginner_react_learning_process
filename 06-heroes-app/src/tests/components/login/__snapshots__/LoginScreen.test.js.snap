// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <LoginComponent /> It should display <LoginComponent /> properly 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
